### PR TITLE
Implement initial sra-report command line tool

### PR DESCRIPTION
**Pull request by BMillman19**
## Description

Command line tool for generating standard relayer API compliance reports

## Motivation and Context

Relayer developers need an easy way to assess their standard relayer API compliance

## How Has This Been Tested?

```bash
cd packages/sra-report
yarn build && node ./lib/index.js -e "https://api.radarrelay.com/0x/v0"
```

## Types of changes

* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

* [ ] Change requires a change to the documentation.
* [ ] Added tests to cover my changes.
* [x] Added new entries to the relevant CHANGELOGs.
* [x] Updated the new versions of the changed packages in the relevant CHANGELOGs.
* [x] Labeled this PR with the 'WIP' label if it is a work in progress.
* [x] Labeled this PR with the labels corresponding to the changed package.

### CODE REVIEW

File: packages/sra-report/postman_configs/environments/kovan_tokens.postman_environment.json [hunk 1]
```diff
@@ -0,0 +1,34 @@
+{
+    "id": "8dab7221-c8b8-2931-92c8-605f6f4072e1",
+    "name": "Kovan Tokens",
+    "values": [
+        {
```
Comment (BMillman19) [inline packages/sra-report/postman_configs/environments/kovan_tokens.postman_environment.json]: remove this

File: packages/sra-report/README.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+This package allows you to generate reports detailing an endpoint's [standard relayer API HTTP specification](https://github.com/0xProject/standard-relayer-api/blob/master/http/v0.md) compliance.
+
+## Installation
+
+`yarn add -g @0xproject/sra-report`
+
+## Usage
+
+```
+abi-gen
```
Comment (fabioberger) [inline packages/sra-report/README.md]: Change to your tool name ;)

File: packages/sra-report/src/types.ts [hunk 3]
```diff
@@ -0,0 +1,3 @@
+export interface Schema {
```
Comment (BMillman19) [inline packages/sra-report/src/types.ts]: realized theres a `Schema` interface exported by our json schemas package, so I'll just use that

File: packages/sra-report/src/globals.d.ts [hunk 4]
```diff
@@ -0,0 +1,6 @@
+declare module 'newman';
```
Comment (fabioberger) [inline packages/sra-report/src/globals.d.ts]: If possible, let's write types for this.

File: packages/sra-report/src/utils.ts [hunk 5]
```diff
@@ -0,0 +1,5 @@
+export const utils = {
+    log(...args: any[]): void {
+        console.log(...args); // tslint:disable-line:no-console
```
Comment (BMillman19) [inline packages/sra-report/src/utils.ts]: we have a couple instances of this pattern across multiple packages, we should consider moving it to the utils package or something

File: packages/sra-report/src/postman_environment_factory.ts [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+import * as mainnetTokensEnvironmentJSON from '../postman_configs/environments/mainnet_tokens.postman_environment.json';
+
+interface EnvironmentValue {
+    key: string;
+    value: string;
+    enabled: boolean;
+    type: string;
+}
+
+export const postmanEnvironmentFactory = {
```
Comment (fabioberger) [inline packages/sra-report/src/postman_environment_factory.ts]: Could you add a comment here explaining what it does. It's not immediately clear to me.

File: packages/sra-report/src/postman_environment_factory.ts [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+        return environment;
+    },
+    createNetworkEnvironment(networkId: number) {
+        switch (networkId) {
+            case 1:
+                return mainnetTokensEnvironmentJSON;
+            case 42:
+                return kovanTokensEnvironmentJSON;
+            default:
+                return {};
```
Comment (fabioberger) [inline packages/sra-report/src/postman_environment_factory.ts]: Why does this return an empty object? Shouldn't it throw when passed an unsupported networkId?

File: packages/sra-report/src/postman_environment_factory.ts [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+import { Schema, schemas as schemasByName } from '@0xproject/json-schemas';
+import * as _ from 'lodash';
+
+import * as kovanTokensEnvironmentJSON from '../postman_configs/environments/kovan_tokens.postman_environment.json';
+import * as mainnetTokensEnvironmentJSON from '../postman_configs/environments/mainnet_tokens.postman_environment.json';
+
+interface EnvironmentValue {
+    key: string;
+    value: string;
+    enabled: boolean;
```
Comment (fabioberger) [inline packages/sra-report/src/postman_environment_factory.ts]: Can you add a comment to the right of this property. What is it enabling?

File: packages/sra-report/src/utils.ts (hunk 5 above)
Comment (fabioberger) [reply packages/sra-report/src/utils.ts]: Yeah, I'm 100% for this.

File: packages/sra-report/tsconfig.json [hunk 9]
```diff
@@ -0,0 +1,11 @@
+{
+    "extends": "../../tsconfig",
+    "compilerOptions": {
+        "outDir": "lib"
+    },
+    "include": [
+        "./src/**/*",
+        "../../node_modules/web3-typescript-typings/index.d.ts",
+        "../../node_modules/ethers-typescript-typings/index.d.ts"
```
Comment (fabioberger) [inline packages/sra-report/tsconfig.json]: Remove both of these. Neither is being used by this package.

Comment (BMillman19) [reply packages/sra-report/tsconfig.json]: We need to look into this, any TS project that has 0x.js as a dependency will not compile w/o them

File: packages/sra-report/src/utils.ts (hunk 5 above)
Comment (BMillman19) [reply packages/sra-report/src/utils.ts]: I'll follow up with a separate PR

File: packages/sra-report/package.json [hunk 10]
```diff
@@ -0,0 +1,45 @@
+{
+    "name": "@0xproject/sra-report",
+    "version": "0.0.1",
+    "description": "Generate reports for standard relayer API compliance",
+    "main": "lib/index.js",
+    "types": "lib/index.d.ts",
+    "scripts": {
+        "build:watch": "tsc -w",
+        "lint": "tslint --project . 'src/**/*.ts'",
```
[the remark this answers is not in the document]
Comment (BMillman19) [reply packages/sra-report/package.json]: not yet

File: packages/sra-report/src/index.ts [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+import { postmanEnvironmentFactory } from './postman_environment_factory';
+import { utils } from './utils';
+
+const newmanRunAsync = promisify<void>(newman.run);
+const DEFAULT_NETWORK_ID = 1;
+const SUPPORTED_NETWORK_IDS = [1, 42];
+
+// extract command line arguments
+const args = yargs
+    .option('url', {
```
[the remark this answers is not in the document]
Comment (BMillman19) [reply packages/sra-report/src/index.ts]: I went with endpoint-url

File: packages/sra-report/src/index.ts [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+              reporters: 'cli',
+          };
+    const environment = !_.isUndefined(args.environment)
+        ? args.environment
+        : await postmanEnvironmentFactory.createPostmanEnvironmentAsync(args.endpointUrl, args.networkId);
+    const newmanRunOptions = {
+        collection: sraReportCollectionJSON,
+        environment,
+        exportCollection: args.exportCollection,
+        exportEnvironment: args.exportEnvironment,
```
Comment (dekz) [inline packages/sra-report/src/index.ts]: This seems to be exporting without the `name` attribute, causing imports into postman to fail.

File: packages/sra-report/src/contract_addresses/mainnet_addresses.ts [hunk 13]
```diff
@@ -0,0 +1,5 @@
+export const addresses = {
+    WETH: '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2',
```
Comment (dekz) [inline packages/sra-report/src/contract_addresses/mainnet_addresses.ts]: Do you think it makes sense to pull these from the ABI? Hardcoded is probably ok, just a bit magic.

Comment (BMillman19) [reply packages/sra-report/src/contract_addresses/mainnet_addresses.ts]: i think its probably ok

**Comment by BMillman19** [comment]
@dekz re: failure cases, writing a proper suite of unit tests for this will include synthesizing many different fixtures for different cases, i'm going to follow up in another PR with all the unit tests, including ones testing the environment creation. I've created all the relevant tasks in the backlog

File: packages/sra-report/package.json (hunk 10 above)
[the remark this answers is not in the document]
Comment (fabioberger) [reply packages/sra-report/package.json]: spoke offline about adding them in a subsequent PR.

File: packages/sra-report/README.md [hunk 14]
```diff
@@ -0,0 +1,105 @@
+# SRA Report
+
+This tool allows you to generate reports that detail an endpoint's [standard relayer API HTTP specification](https://github.com/0xProject/standard-relayer-api/blob/master/http/v0.md) compliance. The tool will perform a [Postman collection](https://www.getpostman.com/docs/v6/postman/collections/creating_collections) run and either print a report to the console or save it to disk as a json file. SRA report can also output a Postman collection file and [Postman environment](https://www.getpostman.com/docs/v6/postman/environments_and_globals/manage_environments) file in order to facilitate replication and debugging of collection runs using the [Postman native app](https://www.getpostman.com/docs/v6/postman/launching_postman/installation_and_updates).
+
```
Comment (fabioberger) [inline packages/sra-report/README.md]: Can we add a bullet list of what the tests are currently checking?

File: packages/sra-report/src/contract_addresses/mainnet_addresses.ts (hunk 13 above)
Comment (fabioberger) [reply packages/sra-report/src/contract_addresses/mainnet_addresses.ts]: Agreed with @BMillman19. Won't change anytime soon.

File: packages/sra-report/src/index.ts [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+import * as yargs from 'yargs';
+
+import * as sraReportCollectionJSON from '../postman_configs/collections/sra_report.postman_collection.json';
+
+import { postmanEnvironmentFactory } from './postman_environment_factory';
+import { utils } from './utils';
+
+const newmanRunAsync = promisify<void>(newman.run);
+const DEFAULT_NETWORK_ID = 1;
+const SUPPORTED_NETWORK_IDS = [1, 42];
```
Comment (fabioberger) [inline packages/sra-report/src/index.ts]: Can we please add support for Rinkeby and Ropsten? We are deployed on those networks and I know for a fact that some relayers are using them as their testnets.

File: packages/sra-report/tsconfig.json (hunk 9 above)
Comment (fabioberger) [reply packages/sra-report/tsconfig.json]: @BMillman19 can you double check that this is still the case? Fixed some related issue over the weekend.

**Comment by fabioberger** [comment]
@BMillman19 please add this package to the top-level README packages table 💃

**Pull request merged**